#!/bin/bash
#PBS -q workq
#PBS -N GWAS_QC
#PBS -V 
#PBS -S /bin/bash 
#PBS -l nodes=1:ppn=6
#PBS -o QC_steps.log
#PBS -e QC_steps.err

cd $PBS_O_WORKDIR

#QC steps according to Visscher paper  (2022)


#input 
input=/home/edizadehm/03.resources/03.SCZ_CC/02.original_SCZ_CC_data/SCZ_CC_autosomes
phenotpic_miss_list=/home/edizadehm/03.resources/03.SCZ_CC/02.original_SCZ_CC_data/missing_phenotypes_individual_list.txt

#outputs
output_dir=/home/edizadehm/03.resources/03.SCZ_CC/03.staging/
base_name=SCZ_CC_3465
base=${output_dir}${base_name}
#
phen_miss=${base}_phen_miss
miss_check=${base}_miss_check
ind_miss=${base}_ind_miss
geno_miss=${base}_geno_miss
maf_filtered=${base}_maf
maf_check=${base}_maf_check
hwe_controls_filtered=${base}_hwe_controls
hwe_all_filtered=${base}_hwe_all
hwe_check=${base}_hwe_check
indepSNP=${base}_indepSNP
het_check=${base}_het_check
het_filtered=${base}_het_filtered
#scripts
plink=/home/edizadehm/01.software/02.plink/plink
script_base=/home/edizadehm/03.resources/03.SCZ_CC/01.scripts/
inversion=${script_base}inversion.txt
maf_check_script=${script_base}MAF_check.R
hist_miss_script=${script_base}hist_miss.R
check_heterozygosity_rate=${script_base}check_heterozygosity_rate.R
heterozygosity_outliers_list=${script_base}heterozygosity_outliers_list.R
hwe_script=${script_base}hwe.R

date

#Delete samples that has no phenotpic data or value
# awk '$6 == 0' SCZ_CC_autosomes.fam | cut -d' ' -f1 > missing_phenotypes_individual_list.txt
$plink --bfile $input --remove $phenotpic_miss_list --make-bed --out $phen_miss

### Step 1 ### 
# Investigate missingness per individual and per SNP and make histograms.

#check missingness
$plink --bfile $phen_miss --missing --out $miss_check 
# output: plink.imiss and plink.lmiss, these files show respectively the proportion of missing SNPs per individual and the proportion of missing individuals per SNP.

# Generate plots to visualize the missingness results.
Rscript --no-save $hist_miss_script $miss_check 

# Delete SNPs with missingness >0.05.
$plink --bfile $phen_miss --geno 0.05 --make-bed --out $ind_miss

# Delete individuals with missingness >0.05.
$plink --bfile $ind_miss --mind 0.05 --make-bed --out $geno_miss



### Step2 ####

# Check for sex discrepancy.

### Step 3 ### 

# Generate a bfile with autosomal SNPs only and delete SNPs with a low minor allele frequency (MAF).
# Select autosomal SNPs only (i.e., from chromosomes 1 to 22).
# awk '{ if ($1 >= 1 && $1 <= 22) print $2 }' HapMap_3_r3_6.bim > snp_1_22.txt
# plink --bfile HapMap_3_r3_6 --extract snp_1_22.txt --make-bed --out HapMap_3_r3_7
# Generate a plot of the MAF distribution.
$plink --bfile $geno_miss --freq --out $maf_check
Rscript --no-save $maf_check_script $maf_check

# Remove SNPs with a low MAF frequency.
$plink --bfile $geno_miss --maf 0.0001 --make-bed --out $maf_filtered



### Step 4 ###

# Delete SNPs which are not in Hardy-Weinberg equilibrium (HWE).
# Check the distribution of HWE p-values of all SNPs.

$plink --bfile $maf_filtered --hardy --out $hwe_check
# Selecting SNPs with HWE p-value below 0.00001, required for one of the two plot generated by the next Rscript, allows to zoom in on strongly deviating SNPs. 
awk '{ if ($9 <0.00001) print $0 }' ${hwe_check}.hwe > ${hwe_check}.plinkzoomhwe.hwe
Rscript --no-save $hwe_script $hwe_check

# By default the --hwe option in plink only filters for controls.
# Therefore, we use two steps, first we use a stringent HWE threshold for controls, followed by a less stringent threshold for the case data.
$plink --bfile $maf_filtered --hwe 1e-6 --make-bed --out $hwe_controls_filtered

# This second HWE step only focusses on cases because in the controls all SNPs with a HWE p-value < hwe 1e-6 were already removed
$plink --bfile $hwe_controls_filtered --hwe 1e-10 --hwe-all --make-bed --out $hwe_all_filtered


### step 5 ###
# Generate a plot of the distribution of the heterozygosity rate of the subjects.
# And remove individuals with a heterozygosity rate deviating more than 3 sd from the mean.
$plink --bfile $hwe_all_filtered --exclude $inversion --range --indep-pairwise 50 5 0.2 --out $indepSNP
$plink --bfile $hwe_all_filtered --extract ${indepSNP}.prune.in --het --out $het_check
# This file contains your pruned data set.

# Plot of the heterozygosity rate distribution
Rscript --no-save $check_heterozygosity_rate $het_check

# The following code generates a list of individuals who deviate more than 3 standard deviations from the heterozygosity rate mean.
Rscript --no-save $heterozygosity_outliers_list $het_check

# Output of the command above: fail-het-qc.txt .
# When using our example data/the HapMap data this list contains 2 individuals (i.e., two individuals have a heterozygosity rate deviating more than 3 SD's from the mean).
# Adapt this file to make it compatible for PLINK, by removing all quotation marks from the file and selecting only the first two columns.
sed 's/"// g' ${het_check}.fail-het-qc.txt | awk '{print$1, $2}'> ${het_check}.het_fail_ind.txt

# Remove heterozygosity rate outliers.
$plink --bfile $hwe_all_filtered --remove ${het_check}.het_fail_ind.txt --make-bed --out $het_filtered

#Cleaning step
#remove all unecessary files:
rm ${ind_miss}* ${phen_miss}* ${geno_miss}* ${hwe_controls_filtered}* ${hwe_all_filtered}*

#PCA analysis. There are 2 steps in the PCA analysis. For common and rare variants.
#Common variants

date
