#!/bin/bash
#PBS -q workq
#PBS -N SC_Geno
#PBS -V 
#PBS -S /bin/bash 
#PBS -l nodes=1:ppn=6
#PBS -o SC_Geno.log
#PBS -e SC_Geno.err

#input 
plink=/home/edizadehm/01.software/02.plink/plink
gcta=/home/edizadehm/02.tutorials/07.gcta/gcta-1.94.1-linux-kernel-3-x86_64/gcta64
input=/home/edizadehm/03.resources/03.SCZ_CC/02.original_SCZ_CC_data/scz_cc_genos.list
hapmap3_snps=/home/edizadehm/03.resources/02.hapmap/02.hapmap3_snps/hapmap3_rs_only.txt
#outputs
output_dir=/home/edizadehm/03.resources/03.SCZ_CC/03.staging/
base_name=SCZ_CC_GCTA_common
base=${output_dir}${base_name}
grm=$base
grm_filtered=${base}_filtered
pca_20=${base}_pca
phen=${base}.phen
output=${base}.final

# PAPER NOTE:
# For the remaining 34,038 individuals, we built a genetic relatedness matrix using variants on
# HM3 reference panels with a MAF > 0.01 and removed one of each pair of individuals with
# estimated genetic relatedness > 0.05, resulting in 28,755 unrelated individuals.
# Before estimating the proportion of phenotypic variance due to additive genetic factors from
# WGS data we initially wanted to check for consistency with previous studies and performed
# a single-component GREML analysis (GREML-SC approach) in GCTA using HM3 SNPs
# with 20 PCs. We calculated the PCs from LD-pruned SNPs and fitted them as fixed effects
# in the GREML-SC analysis.

#I am going to use multiple geno files instead of plink ped format.
#The geno files here are not filtered for imputation score, there are just 
# unfiltered, raw genotype probabilities for all of the samples. 

#multigeno file: include all of the geno files pathes.

#make grm
bgen=/home/edizadehm/03.resources/03.SCZ_CC/02.original_SCZ_CC_data/genos/SCZ_CC_chr.gen
sample=/home/edizadehm/03.resources/03.SCZ_CC/02.original_SCZ_CC_data/genos/SCZ_CC_chr.sample
$gcta --bgen  $bgen --sample $sample --autosome --maf 0.01 --make-grm --out $grm --thread-num 10
echo -e "+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n\n" 

# #removing cryptic relatedness 
# $gcta --grm $grm --grm-cutoff 0.05  --make-grm  --out $grm_filtered
# echo -e "+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n\n"

# $gcta  --grm $grm_filtered --keep ${grm_filtered}.grm.id --extract $hapmap3_snps --pca 20  --out $pca_20
# echo -e "+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n\n"

# #Creating Phen files from the 
# #For a case-control study, the phenotypic values of cases and controls should be specified as 1 and 0, respectively. Suppose you have prepared a phenotype file test_cc.phen. 
# cat ${input}.fam | awk '{if ($6 == 1) $6 = 0; else if ($6 == 2) $6 = 1; print $1, $2, $6}' > $phen
# echo -e "+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n\n"
# # GREML
# $gcta --grm $grm_filtered --qcovar ${pca_20}.eigenvec --pheno $phen --reml --prevalence 0.01 --out $output --thread-num 10
# echo -e "+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n\n"